src/main/java/org/roaringbitmap/longlong/RoaringTreeMap.java

```diff
@@ -0,0 +1,750 @@
+
+    sortedCumulatedCardinality = new long[0];
+    sortedHighs = new int[0];
+
+    // In case of de-serialization, we need to add back all bitmaps
+    lowBitmaps.clear();
+    // lowBitmaps.addAll(highToBitmap.values());
+
+    latestAddedHigh = null;
+  }
+
+  /**
+   * Add the value to the container (set the value to "true"), whether it already appears or not.
+   *
+   * Java lacks native unsigned integers but the x argument is considered to be unsigned. Within
+   * bitmaps, numbers are ordered according to {@link Integer#compareUnsigned}. We order the numbers
+   * like 0, 1, ..., 2147483647, -2147483648, -2147483647,..., -1.
+   *
+   * @param x integer value
+   */
+  public void addLong(long x) {
+    int high = high(x);
+    int low = low(x);
+
+    // Copy the reference to prevent race-condition
+    Map.Entry<Integer, BitmapDataProvider> local = latestAddedHigh;
+
+    BitmapDataProvider bitmap;
+    if (local != null && local.getKey().intValue() == high) {
+      bitmap = local.getValue();
+    } else {
+      bitmap = highToBitmap.get(high);
+      if (bitmap == null) {
+        bitmap = new MutableRoaringBitmap();
+        pushBitmapForHigh(high, bitmap);
+      }
+      latestAddedHigh = new AbstractMap.SimpleImmutableEntry<>(high, bitmap);
+    }
+    bitmap.add(low);
+
+    // The cardinalities after this bucket may not be valid anymore
+    if (compare(firstHighNotValid, high) > 0) {
+      firstHighNotValid = high;
+
+      int indexNotValid = binarySearch(sortedHighs, firstHighNotValid);
+
+      final int indexAfterWhichToReset;
+      if (indexNotValid >= 0) {
+        indexAfterWhichToReset = indexNotValid;
+      } else {
+        // We have added a value for a branch new high
+        indexAfterWhichToReset = -indexNotValid - 1;
+      }
+
+      // This way, sortedHighs remains sorted
+      Arrays.fill(sortedHighs, indexAfterWhichToReset, sortedHighs.length, highestHigh());
+    }
+    allValid = false;
+  }
+
+  private int compare(int x, int y) {
+    if (signedLongs) {
+      return Integer.compare(x, y);
+    } else {
+      return RoaringIntPacking.compareUnsigned(x, y);
+    }
+  }
+
+  private void pushBitmapForHigh(int high, BitmapDataProvider bitmap) {
+    // TODO .size is too slow
+    // int nbHighBefore = highToBitmap.headMap(high).size();
+
+    BitmapDataProvider previous = highToBitmap.put(high, bitmap);
+    if (previous != null) {
+      throw new IllegalStateException("Should push only not-existing high");
+    }
+
+    // If there is 1 bucket before, we need to add at index 1
+    // TODO .size were too slow
+    // lowBitmaps.add(nbHighBefore, bitmap);
+  }
+
+  private int low(long id) {
+    return RoaringIntPacking.low(id);
+  }
+
+  private int high(long id) {
+    return RoaringIntPacking.high(id);
+  }
+
+  /**
+   * Returns the number of distinct integers added to the bitmap (e.g., number of bits set).
+   *
+   * @return the cardinality
+   */
+  public long getLongCardinality() {
```

Comment:
I tried adding to confirm we can have a cardinality above (2L * Integer.MAX_VALUE + 3L), the first cardinality rejected by RoaringBitmap. But something seems wrong with Roaring64NavigableMap.add(long,long).

I'm unsure how to document properly the fact the cardinality is bounded by Integer.MAX_VALUE * RoaringBitmap cardinality. The TreeMap may even be able to grow to above Integer.MAX_VALUE...